Implement `is (not) defined` feature

GuillaumeGomez opened the pull request:
Fixes https://github.com/rinja-rs/rinja/issues/62.

I'm not super happy about how I implemented the parser part (the little "cheat") but found it easier to be done this way to keep the same logic as the other operators. But it forced to have some more error checking.

The other downside is that it makes the code to generate the `if`/`else` branches a bit messy. If you have ideas on how to make it look better by any chance?

Kijewski commented:
Sorry, but I won't find the time to review before Friday/Saturday.

Kijewski commented on rinja_parser/src/expr.rs:
```diff
@@ -188,7 +190,59 @@ impl<'a> Expr<'a> {
+        let (i, left) = Self::filtered(i, level)?;
+        // This is cheating: to make it easier, we expect `defined` to be provided afterward
+        // in order to keep the same code logic as the other calls.
+        let (i, right) = many0(pair(tuple((ws(tag("is ")), ws(opt(tag("not "))))), |i| {
```
```rs
let (i, is_neg) = preceded(
    ws(keyword("is")),
    cut(terminated(
        opt(keyword("not")),
        ws(keyword("defined")),
    )),
)(i)?;
```

GuillaumeGomez commented:
> Sorry, but I won't find the time to review before Friday/Saturday.

You say that and then provide suggestion right away. 😆 

But in any case, it's not in a hurry. So please review when you have time and motivation to keep this project fun for you. :)

GuillaumeGomez commented:
Finally applied your suggestion for the parser. Looks much better like this, thanks!

GuillaumeGomez commented:
> The other downside is that it makes the code to generate the `if`/`else` branches a bit messy. If you have ideas on how to make it look better by any chance?

Thinking some more about this, I think we can simply create a new very small AST for conditions which would simply remove the conditions (and their associated block) that won't be rendered and then the function can simply loop through them instead of the whole state machine. Gonna do that.

GuillaumeGomez commented:
Moved the logic out of the `write_if` method. Much closer to what it originally was.

Kijewski commented:
> Moved the logic out of the `write_if` method. Much closer to what it originally was.

Thank you, much easier to read! I give you a proper review soon-ish.

Kijewski commented on rinja_parser/src/expr.rs:
```diff
@@ -188,7 +190,56 @@ impl<'a> Expr<'a> {
+    expr_prec_layer!(muldivmod, is_defined, alt((tag("*"), tag("/"), tag("%"))));
+
+    fn is_defined(i: &'a str, level: Level) -> ParseResult<'a, WithSpan<'a, Self>> {
+        let (_, level) = level.nest(i)?;
```
I think we can to without the nesting check in here. We only traverse to a lower level, not the same or a higher level.

GuillaumeGomez replied:
Don't we need that for `Self::filtered(i, level)?` just below?

Kijewski replied:
No, you can just pass the `level` as is. In all other cases we also only increment the recursion level on recursions. The code cannot parse an expression like `y is defined is defined`, so no nesting happens in here.

GuillaumeGomez replied:
Good point! Removed this nesting.

Kijewski commented on rinja_parser/src/expr.rs:
```diff
@@ -188,7 +190,56 @@ impl<'a> Expr<'a> {
+            Some(Some(Some(_))) => true,
+        };
+        match *lhs {
+            Self::Var(_) => {}
```
Please capture the variable name in here.

GuillaumeGomez replied:
Done!

Kijewski commented on rinja_parser/src/expr.rs:
```diff
@@ -188,7 +190,56 @@ impl<'a> Expr<'a> {
+            i,
+            WithSpan::new(
+                if is_neg {
+                    Self::IsNotDefined(Box::new(lhs))
```
The use the variable name in here. No need for boxing. This makes the code a lot more simple, I think.

GuillaumeGomez replied:
That was an incredible idea! Code got simplified by quite a lot! :D

Kijewski commented:
Sorry for the iterative reviews! It's a bit hard for me to wrap my head around this feature.

GuillaumeGomez commented:
> Sorry for the iterative reviews! It's a bit hard for me to wrap my head around this feature.

It's fine. Don't hesitate if you need some clarifications or any help to understand the code (which, if not easy to understand, should be more documented! So don't hesitate to tell me what's unclear so I can add code comments :smiley:)

Kijewski commented:
I pushed a commit to remove the line of unsafe code. I actually wanted to look if it was easily possible to remove it, and the best test was implementing it. :D

The code looks good, and after reading your comments properly, I understand it perfectly (I think). Great work!

GuillaumeGomez commented:
Fixed the merge conflict and applied your suggestion. :)

> I pushed a commit to remove the line of unsafe code. I actually wanted to look if it was easily possible to remove it, and the best test was implementing it. :D

That's my first implementation. I removed it to reduce the number of lifetimes. If you prefer with references, then I'm fine with it. :)

> The code looks good, and after reading your comments properly, I understand it perfectly (I think). Great work!

Awesome! \o/

Kijewski reviewed (dismissed):
Thank you! Please revert the changes in the ui test nesting tests, then you can merge the PR.

GuillaumeGomez commented:
Reverted wrong nested ui tests changes.

Kijewski reviewed (approved):
Thank you, well done!
